### PR TITLE
[KernelGen][Nvidia] Fix _log_softmax

### PR DESCRIPTION
**Pull request by Yukun-Cui**
## Summary
Fixes the `log_softmax` operator.

## Testing
- Parametrized tests over `REDUCTION_SHAPES`, `FLOAT_DTYPES` and `dim`
- Validated against reference on device via `to_reference(inp, True)`
- Local run: `pytest tests/test_log_softmax.py` → 36 passed (0 warnings)
- Local run: `pytest benchmark/test_log_softmax.py --level core` → 4 passed
- Tested on: Nvidia (H20)

## Performance
Test command: `pytest benchmark/test_log_softmax.py --level core` (NVIDIA H20)

### log_softmax

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|-------|------|--------------------|-------------------|---------|--------|
| float16 | (64, 64) | 5.7920 | 6.7840 | 0.854 | — |
| float16 | (256, 256) | 6.4960 | 7.0400 | 0.923 | — |
| float16 | (1024, 1024) | 9.6000 | 11.9680 | 0.802 | — |
| float32 | (64, 64) | 6.2080 | 6.7840 | 0.915 | — |
| float32 | (256, 256) | 6.5920 | 7.0720 | 0.932 | — |
| float32 | (1024, 1024) | 9.2160 | 12.0000 | 0.768 | — |
| bfloat16 | (64, 64) | 5.6640 | 6.7840 | 0.835 | — |
| bfloat16 | (256, 256) | 6.8480 | 7.0400 | 0.973 | — |
| bfloat16 | (1024, 1024) | 9.8880 | 11.9360 | 0.828 | — |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| log_softmax | **0.870** |

### log_softmax_out

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|-------|------|--------------------|-------------------|---------|--------|
| float16 | (64, 64) | 5.8560 | 7.1040 | 0.824 | — |
| float16 | (1024, 1024) | 9.8560 | 11.6160 | 0.848 | — |
| float16 | (4096, 4096) | 81.1200 | 61.6960 | 1.315 | — |
| float16 | (64, 512, 512) | 484.7360 | 68.9280 | 7.032 | — |
| float16 | (1024, 1024, 1024) | 8738.4644 | 2271.4721 | 3.847 | — |
| float32 | (64, 64) | 6.2080 | 6.5600 | 0.946 | — |
| float32 | (1024, 1024) | 9.5040 | 12.6080 | 0.754 | — |
| float32 | (4096, 4096) | 81.4560 | 76.6400 | 1.063 | — |
| float32 | (64, 512, 512) | 601.8240 | 88.1600 | 6.826 | — |
| float32 | (1024, 1024, 1024) | 9375.5040 | 3575.5200 | 2.622 | — |
| bfloat16 | (64, 64) | 6.2720 | 7.0720 | 0.887 | — |
| bfloat16 | (1024, 1024) | 9.9520 | 11.5840 | 0.859 | — |
| bfloat16 | (4096, 4096) | 80.6400 | 62.4000 | 1.292 | — |
| bfloat16 | (64, 512, 512) | 482.4640 | 77.1200 | 6.256 | — |
| bfloat16 | (1024, 1024, 1024) | 8729.0878 | 2487.8080 | 3.509 | — |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| log_softmax_out | **2.592** |

### log_softmax_backward_data

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|-------|------|--------------------|-------------------|---------|--------|
| float16 | (64, 64) | 5.7600 | 6.3360 | 0.909 | — |
| float16 | (256, 256) | 6.6560 | 6.8800 | 0.967 | — |
| float16 | (1024, 1024) | 11.2960 | 12.1280 | 0.931 | — |
| float16 | (4096, 4096) | 38.7840 | 71.4880 | 0.543 | — |
| float16 | (1024, 65536) | 154.2080 | 396.1600 | 0.389 | — |
| float32 | (64, 64) | 6.1120 | 6.3040 | 0.970 | — |
| float32 | (256, 256) | 6.5920 | 7.5840 | 0.869 | — |
| float32 | (1024, 1024) | 10.9120 | 12.0960 | 0.902 | — |
| float32 | (4096, 4096) | 68.1600 | 86.2400 | 0.790 | — |
| float32 | (1024, 65536) | 357.0240 | 430.1600 | 0.830 | — |
| bfloat16 | (64, 64) | 5.7600 | 6.4640 | 0.891 | — |
| bfloat16 | (256, 256) | 7.1040 | 6.8480 | 1.037 | — |
| bfloat16 | (1024, 1024) | 11.6160 | 12.3520 | 0.940 | — |
| bfloat16 | (4096, 4096) | 39.9680 | 71.7440 | 0.557 | — |
| bfloat16 | (1024, 65536) | 154.3680 | 395.2320 | 0.391 | — |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| log_softmax_backward_data | **0.794** |

### log_softmax_backward_data_out

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|-------|------|--------------------|-------------------|---------|--------|
| float16 | (64, 64) | 6.2080 | 6.4960 | 0.956 | — |
| float16 | (256, 256) | 6.8480 | 6.7520 | 1.014 | — |
| float16 | (1024, 1024) | 11.6800 | 12.0640 | 0.968 | — |
| float16 | (4096, 4096) | 38.6720 | 72.9600 | 0.530 | — |
| float16 | (1024, 65536) | 153.9040 | 396.3840 | 0.388 | — |
| float32 | (64, 64) | 5.7280 | 6.2720 | 0.913 | — |
| float32 | (256, 256) | 6.6240 | 7.8080 | 0.848 | — |
| float32 | (1024, 1024) | 11.1360 | 12.0640 | 0.923 | — |
| float32 | (4096, 4096) | 67.9360 | 85.5040 | 0.795 | — |
| float32 | (1024, 65536) | 386.1760 | 433.2800 | 0.891 | — |
| bfloat16 | (64, 64) | 6.1760 | 6.2720 | 0.985 | — |
| bfloat16 | (256, 256) | 6.8480 | 6.9760 | 0.982 | — |
| bfloat16 | (1024, 1024) | 10.9440 | 12.2880 | 0.891 | — |
| bfloat16 | (4096, 4096) | 39.9680 | 72.3840 | 0.552 | — |
| bfloat16 | (1024, 65536) | 155.3920 | 395.3280 | 0.393 | — |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| log_softmax_backward_data_out | **0.802** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (36 cases) | 0.870/2.592/0.794/0.802 | Primary |
| Tianshu | N/A | — | Not tested locally |
| Muxi | N/A | — | Not tested locally |
| Ascend | N/A | — | Not tested locally |
| Hygon | N/A | — | Not tested locally |

## Files Changed
- `src/flag_gems/ops/log_softmax.py`: Replace deprecated `and` with `&` on tensor masks (4 sites) with explanatory comment
- `tests/test_log_softmax.py`: Remove `use_gems()`, call `flag_gems.log_softmax*` wrappers directly; use `torch._log_softmax` for the reference
- `benchmark/test_log_softmax.py`: Replace `torch.nn.functional.log_softmax` with `torch._log_softmax` (with `dim`/`half_to_float` args)

### CODE REVIEW

File: src/flag_gems/ops/log_softmax.py [hunk 1]
```diff
@@ -150,7 +152,8 @@ def log_softmax_backward_kernel(
     for start_n in range(0, N, BLOCK_N):
         n_offset = start_n + tl.arange(0, BLOCK_N)
         offsets = m_offset[:, None] * N * K + n_offset[None, :] * K + pid_k
-        mask = m_offset[:, None] < M and n_offset[None, :] < N
+        # '&' for element-wise tensor AND; Python 'and' is deprecated on tensors
+        mask = (m_offset[:, None] < M) & (n_offset[None, :] < N)
```
Comment (huangyiqun) [inline src/flag_gems/ops/log_softmax.py:156]: The masked load below feeds a reduction but has no other value. Tail lanes are undefined when N is not a multiple of BLOCK_N; please use other=0.0.

Comment (Yukun-Cui) [reply src/flag_gems/ops/log_softmax.py:156]: Good catch! I'll add `other=0.0` to the masked load to ensure tail lanes don't contribute undefined values to the reduction when N is not a multiple of BLOCK_N.
